core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -924,25 +900,33 @@ private object RTS {
       io.ensuring(exitUninterruptible)
     }
 
-    final def register(cb: Callback[E, A]): Async[E, A] = join0(cb)
+    final def register(cb: Callback[E, A]): Async[E, A] =
+      observe0 {
+        case ExitResult.Completed(r) => cb(r)
+        case _                       =>
+      } match {
+        case Async.Now(v)          => Async.Now(v.fold(identity, ExitResult.Failed(_, _), ExitResult.Terminated(_)))
```

Comment:
This match block looks like it could be replaced by adding some map function to `Async`.